### PR TITLE
Windows Tests

### PR DESCRIPTION
**Pull request by varunagrawal**
Thanks to the awesome work by @mikesheffler, I was finally able to figure out what was the issue behind all the "undefined" errors.

This PR adds/removes all the necessary `GTSAM_EXPORT` declarations and updates the CI to run all `GTSAM` test targets. Unfortunately some targets have test failures due to bugs creeping in over time and the CI not being enabled for them, but those can be tackled later.

I am leaving `GTSAM_UNSTABLE` test targets for another day.

### CODE REVIEW

File: gtsam/geometry/Line3.h [hunk 1]
```diff
@@ -146,7 +146,30 @@ class Line3 {
  */
 Line3 transformTo(const Pose3 &wTc, const Line3 &wL,
                   OptionalJacobian<4, 6> Dpose = boost::none,
-                  OptionalJacobian<4, 4> Dline = boost::none);
+                  OptionalJacobian<4, 4> Dline = boost::none) {
```
Comment (dellaert) [inline gtsam/geometry/Line3.h:149]: I had trouble with definitions in headers. Should be inline? But I don't understand why we can't just export here and leave definition in .cpp?

File: gtsam/navigation/tests/testSerializationNavigation.cpp [hunk 2]
```diff
@@ -39,6 +39,9 @@ BOOST_CLASS_EXPORT_GUID(noiseModel::Unit, "gtsam_noiseModel_Unit")
 BOOST_CLASS_EXPORT_GUID(noiseModel::Isotropic, "gtsam_noiseModel_Isotropic")
 BOOST_CLASS_EXPORT_GUID(SharedNoiseModel, "gtsam_SharedNoiseModel")
 BOOST_CLASS_EXPORT_GUID(SharedDiagonal, "gtsam_SharedDiagonal")
+BOOST_CLASS_EXPORT_GUID(PreintegratedImuMeasurements, "gtsam_PreintegratedImuMeasurements")
```
Comment (dellaert) [inline gtsam/navigation/tests/testSerializationNavigation.cpp:42]: should be gtsam::PreintegratedImuMeasurements I think, not _.
We should really standardize (different issue, different PR, different contributor)

**Review by dellaert — APPROVED** [review]
Awesome (if windows CI actually passes :-))

**Comment by varunagrawal** [comment]
I forgot to comment out the gtsam_unstable test build. Windows CI will fail for that reason.

**Review by dellaert — CHANGES_REQUESTED** [review]
Changing to "request changes" as I think that function should not be moved out of cpp, unless you can tell me the reason :-)

**Comment by varunagrawal** [comment]
The linker seems to confuse the transformTo method for Line3 with the expressions version, this throwing a linker error. Another solution would be to rename the expression version to `transformTo_` but that would be API breaking?

**Comment by mikesheffler** [comment]
Just as a datapoint, I currently have (for `fix/windows-tests`, and without attempting any `gtsam_unstable` stuff):

Pass
-----
`check.base`
`check.basis`
`check.inference`
`check.navigation`
`check.sfm`
`check.symbolic`

Fail
-----
`check.discrete`
`check.geometry`
`check.nonlinear`
`check.partition`
`check.sam`
`check.slam`
`check.tests`

I thought `check.discrete` was supposed to be in the 'Pass' category, so I'll print what I have on my side. I should be able to look at this later tonight.

Possibly relevant warning:
-----
```c++
EXPECT(adds = 54);
```
results in

`Assignment '=' used when equality '==' probably intended`

at `testAlgebraicDecisionTree.cpp` line 321.

`check.discrete` Build Errors
-----
Looks like multiply-defined symbols:
```
Build started...
1>------ Build started: Project: check_discrete_program, Configuration: Debug x64 ------
1>gtsamDebug.lib(gtsamDebug.dll) : error LNK2005: "private: static class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl gtsam::AlgebraicDecisionTree<unsigned __int64>::DefaultFormatter(unsigned __int64 const &)" (?DefaultFormatter@?$AlgebraicDecisionTree@_K@gtsam@@CA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEB_K@Z) already defined in testAlgebraicDecisionTree.obj
1>gtsamDebug.lib(gtsamDebug.dll) : error LNK2005: "public: __cdecl gtsam::AlgebraicDecisionTree<unsigned __int64>::AlgebraicDecisionTree<unsigned __int64>(class std::vector<struct std::pair<unsigned __int64,unsigned __int64>,class std::allocator<struct std::pair<unsigned __int64,unsigned __int64> > > const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (??0?$AlgebraicDecisionTree@_K@gtsam@@QEAA@AEBV?$vector@U?$pair@_K_K@std@@V?$allocator@U?$pair@_K_K@std@@@2@@std@@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@3@@Z) already defined in testAlgebraicDecisionTree.obj
1>gtsamDebug.lib(gtsamDebug.dll) : error LNK2005: "public: void __cdecl gtsam::AlgebraicDecisionTree<unsigned __int64>::print(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::function<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl(unsigned __int64)> const &)const " (?print@?$AlgebraicDecisionTree@_K@gtsam@@QEBAXAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBV?$function@$$A6A?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@_K@Z@4@@Z) already defined in testAlgebraicDecisionTree.obj
1>gtsamDebug.lib(gtsamDebug.dll) : error LNK2005: "public: bool __cdecl gtsam::AlgebraicDecisionTree<unsigned __int64>::equals(class gtsam::AlgebraicDecisionTree<unsigned __int64> const &,double)const " (?equals@?$AlgebraicDecisionTree@_K@gtsam@@QEBA_NAEBV12@N@Z) already defined in testAlgebraicDecisionTree.obj
1>gtsamDebug.lib(gtsamDebug.dll) : error LNK2005: "public: virtual __cdecl gtsam::AlgebraicDecisionTree<unsigned __int64>::~AlgebraicDecisionTree<unsigned __int64>(void)" (??1?$AlgebraicDecisionTree@_K@gtsam@@UEAA@XZ) already defined in testAlgebraicDecisionTree.obj
1>gtsamDebug.lib(gtsamDebug.dll) : error LNK2005: "public: __cdecl gtsam::AlgebraicDecisionTree<unsigned __int64>::AlgebraicDecisionTree<unsigned __int64>(class gtsam::AlgebraicDecisionTree<unsigned __int64> const &)" (??0?$AlgebraicDecisionTree@_K@gtsam@@QEAA@AEBV01@@Z) already defined in testAlgebraicDecisionTree.obj
1>gtsamDebug.lib(gtsamDebug.dll) : error LNK2005: "public: __cdecl gtsam::AlgebraicDecisionTree<unsigned __int64>::AlgebraicDecisionTree<unsigned __int64>(struct std::pair<unsigned __int64,unsigned __int64> const &,double,double)" (??0?$AlgebraicDecisionTree@_K@gtsam@@QEAA@AEBU?$pair@_K_K@std@@NN@Z) already defined in testAlgebraicDecisionTree.obj
1>gtsamDebug.lib(gtsamDebug.dll) : error LNK2005: "public: __cdecl gtsam::AlgebraicDecisionTree<unsigned __int64>::AlgebraicDecisionTree<unsigned __int64>(class std::vector<struct std::pair<unsigned __int64,unsigned __int64>,class std::allocator<struct std::pair<unsigned __int64,unsigned __int64> > > const &,class std::vector<double,class std::allocator<double> > const &)" (??0?$AlgebraicDecisionTree@_K@gtsam@@QEAA@AEBV?$vector@U?$pair@_K_K@std@@V?$allocator@U?$pair@_K_K@std@@@2@@std@@AEBV?$vector@NV?$allocator@N@std@@@3@@Z) already defined in testAlgebraicDecisionTree.obj
1>C:\Mike\Work\Build\gtsam\bin\Debug\check_discrete_program.exe : fatal error LNK1169: one or more multiply defined symbols found
1>Done building project "check_discrete_program.vcxproj" -- FAILED.
2>------ Build started: Project: check.discrete, Configuration: Debug x64 ------
2>Test project C:/Mike/Work/Build/gtsam/gtsam/discrete/tests
2>    Start 1: check_discrete_program
2>Could not find executable C:/Mike/Work/Build/gtsam/bin/Debug/check_discrete_program.exe
2>Looked in the following places:
2>C:/Mike/Work/Build/gtsam/bin/Debug/check_discrete_program.exe
2>C:/Mike/Work/Build/gtsam/bin/Debug/check_discrete_program.exe.exe
2>C:/Mike/Work/Build/gtsam/bin/Debug/Debug/check_discrete_program.exe
2>C:/Mike/Work/Build/gtsam/bin/Debug/Debug/check_discrete_program.exe.exe
2>Debug/C:/Mike/Work/Build/gtsam/bin/Debug/check_discrete_program.exe
2>Debug/C:/Mike/Work/Build/gtsam/bin/Debug/check_discrete_program.exe.exe
2>Unable to find executable: C:/Mike/Work/Build/gtsam/bin/Debug/check_discrete_program.exe
2>1/1 Test #1: check_discrete_program ...........***Not Run   0.00 sec
2>
2>0% tests passed, 1 tests failed out of 1
```

**Comment by dellaert** [comment]
Definitely should be ==

 About ADT: could be template instantiation.

**Comment by mikesheffler** [comment]
> About ADT: could be template instantiation.

I think so? As the code says, it (ADT) mostly exists to specialize one template argument to `DecisionTree` as `double`.

I tried sprinkling `inline` here and there in the ADT header on the methods that were complaining about already being defined, but it didn't seem to do anything. Between the `static` methods, the `inline`s and the inheritance, I can't seem to follow it all at once.

Basically, as soon as you have two `ADT` instances in `testAlgebraicDecisionTree.cpp`, you'll get the whole raft of errors, so it's probably something about the whole class. Dunno. 😦

**Comment by dellaert** [comment]
Actually, @mikesheffler , I think the == should be replaced with LONGS_EQUAL(54, adds), and 54 should be changed to the right value ;-)

I think the solution to the ADT linking problem might be to re-add the export, and add a .cpp file with an explicit instantiation on Key. That should create one set of symbols exported to the DLL.

**Comment by mikesheffler** [comment]
**tl;dr**: `check.discrete` passes now.

-----
> I think the == should be replaced with LONGS_EQUAL(54, adds), and 54 should be changed to the right value ;-)

I made the change to `LONGS_EQUAL` and it looks like the value _that the test wants_ is `49`, so I changed it to `49`. I haven't made the slightest attempt to understand what's actually going on in the test, so please take a look and decide if that is sensible.

> the ADT linking problem might be to re-add the export, and add a .cpp file with an explicit instantiation on Key.

I added back `GTSAM_EXPORT` and then made the .cpp with just
```c++
namespace gtsam {

  template class AlgebraicDecisionTree<Key>;

}  // namespace gtsam
```
which seems to convince the linker to export the symbols, and just _once_ to boot. I don't have a tight handle on the linker's decision making process here, so it's worth taking a look.

**Comment by mikesheffler** [comment]
I do not have outstanding changes. I think the idea to chop up the effort makes sense, or this will be unmerged for a while.

**Review by mikesheffler — APPROVED** [review]

**Comment by dellaert** [comment]
Cool. @varunagrawal merge when you're ready
